apstra/api_custom_collector.go

```diff
@@ -0,0 +1,257 @@
+			return Response.Items, nil
+		}
+		return nil, err
+	}
+
+	for i := range Response.Items {
+		Response.Items[i].ServiceName = name
+	}
+	return Response.Items, nil
+}
+
+// CreateCollector creates a collector
+func (o *Client) CreateCollector(ctx context.Context, in *Collector) error {
+	// Check if this is the first collector for that service name
+	cs, err := o.GetCollectorsByServiceName(ctx, in.ServiceName)
+	if err != nil {
+		return err
+	}
+	var Request struct {
+		ServiceName string      `json:"service_name"`
+		Items       []Collector `json:"collectors"`
+	}
+	Request.ServiceName = in.ServiceName
+	Request.Items = append(Request.Items, *in)
+	// This is the first collector for this service name
+	// So we POST
+	if len(cs) == 0 {
+		err = o.talkToApstra(ctx, &talkToApstraIn{
+			method:   http.MethodPost,
+			urlStr:   fmt.Sprintf(apiUrlCollectors),
+			apiInput: &Request,
+		})
+		return err
+	}
+
+	// There are other collectors, so this is a patch
+	err = o.talkToApstra(ctx, &talkToApstraIn{
+		method:   http.MethodPatch,
+		urlStr:   fmt.Sprintf(apiUrlCollectorsByServiceName, in.ServiceName),
+		apiInput: &Request,
+	})
+	return err
+}
+
+// Update Collector Updates a collector
+func (o *Client) UpdateCollector(ctx context.Context, in *Collector) error {
+	var Request struct {
+		Items []Collector `json:"collectors"`
+	}
+	Request.Items = append(Request.Items, *in)
+	return o.talkToApstra(ctx, &talkToApstraIn{
+		method:   http.MethodPatch,
+		urlStr:   fmt.Sprintf(apiUrlCollectorsByServiceName, in.ServiceName),
+		apiInput: &Request,
+	})
+}
+
+// DeleteAllCollectorsInService deletes all the collectors under a service
+func (o *Client) DeleteAllCollectorsInService(ctx context.Context, name string) error {
+	return o.talkToApstra(ctx, &talkToApstraIn{
+		method: http.MethodDelete,
+		urlStr: fmt.Sprintf(apiUrlCollectorsByServiceName, name),
+	})
+}
+
+func (p1 *CollectorPlatform) Equals(p2 *CollectorPlatform) bool {
+	if p1.OsType != p2.OsType {
+		return false
+	}
+	if p1.OsVersion != p2.OsVersion {
+		return false
+	}
+	if p1.Model != p2.Model {
+		return false
+	}
+	if len(p1.OsFamily) != len(p2.OsFamily) {
+		return false
+	}
+
+	m := make(map[CollectorOSVariant]bool, len(p1.OsFamily))
+	for _, v := range p1.OsFamily {
+		m[v] = true
+	}
+	for _, v := range p2.OsFamily {
+		_, ok := m[v]
+		if !ok {
+			return false
+		}
+	}
+	return true
+}
+
+// DeleteCollector deletes the collector described in the object
+func (o *Client) DeleteCollector(ctx context.Context, in *Collector) error {
+	var Request struct {
+		ServiceName string      `json:"service_name"`
+		Items       []Collector `json:"collectors"`
+	}
+
+	cs, err := o.GetCollectorsByServiceName(ctx, in.ServiceName)
+	if err != nil {
+		return err
+	}
+
+	Request.ServiceName = in.ServiceName
+	for _, c := range cs {
+		if !c.Platform.Equals(&in.Platform) {
+			Request.Items = append(Request.Items, c)
+		}
+	}
+	return o.talkToApstra(ctx, &talkToApstraIn{
```

Comment:
needs `convertTtaeToAceWherePossible()`.

Also, it's wild that we have to jump through these hoops.

Do I understand correctly that we have to collect all of the collectors, edit out the target and put the list back to the API?

Seems like another race condition.

Maybe add:

```go
o.lock("telemetry_collector")
defer o.unlock("telemetry_collector")
```

Probably the same required if we're fetching the list elsewhere in order to make changes.

Thoughts?

Reply:
I created a service registry level lock. I thought about a per-collector lock, but concerned that it will leave too many locks lying around.